sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala

```diff
@@ -347,44 +347,81 @@ object GpuCast extends Arm {
 
       // ansi cast from larger-than-integer integral-like types, to integer
       case (t @(LongType | _: DecimalType), IntegerType) if ansiMode =>
-        assertValuesInRange(input, GpuScalar.from(Int.MinValue, t),
-          GpuScalar.from(Int.MaxValue, t))
+        t match {
+          case LongType =>
+            assertValuesInRange[Long](input, Int.MinValue.toLong, Int.MaxValue.toLong)
+          case _ =>
+            assertValuesInRange[BigDecimal](input, BigDecimal(Int.MinValue),
+              BigDecimal(Int.MaxValue))
+        }
         input.castTo(GpuColumnVector.getNonNestedRapidsType(toDataType))
 
       // ansi cast from larger-than-short integral-like types, to short
       case (t @(LongType | IntegerType | _: DecimalType), ShortType) if ansiMode =>
-        assertValuesInRange(input, GpuScalar.from(Short.MinValue, t),
-          GpuScalar.from(Short.MaxValue, t))
+        t match {
```

Comment:
`t` in this `case` (and one more `case` below) can be removed. We can just use `fromDataType`.